### PR TITLE
services/horizon: "Accounts For Signer" endpoint

### LINKED ISSUE #1486: Create new action presenting data from a database

### PR DESCRIPTION
**Pull request by bartekn**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [x] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [x] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### Summary

This PR adds "Accounts For Signer" endpoint that would be hard or impossible to implement without the new ingestion package (`exp/ingest`).

### Goal and scope

The goal of this PR is to add a new endpoint that presents "Accounts For Signer" data (close #1486) and integrate all features listed in #1472 to actually process the data and expose it. This required us to add a new `expingest` package to Horizon that is completely independent of the old `ingest` package and is using the new `exp/ingest` system. The pipeline currently saves signers data to the DB but can be easily extended with more features in the future.

The new endpoint (`/accounts`) is future-proof. It currently works only with `signer` GET parameter set but when more processors are added to the ingestion pipeline we can add new params that will allow filtering accounts by assets, by home domain, by inflation destination etc. When the rest of Horizon is migrated to the new system, `/accounts` will be a collection of Account Details resources (like in `/accounts/{id}`).

### Summary of changes

- Added `/accounts` endpoint.
- Created new `services/horizon/internal/expingest` package. Moved pipelines from `horizon-demo` and updated the processors to process master key changes. Added some extra methods to `xdr` package to support processing ledger entry changes.
- Added `key_value_store` table and helper methods. This is to support storing the sequence of the last ledger ingested by the new system.

### Known limitations & issues

- Enabling this feature requires around 4GB of RAM for processing initial state and a space for around 3M rows in `accounts_signers` table for pubnet.
- The endpoint will not be available until the state from history archives is not fully parsed. This usually takes a couple minutes. Before then we display `StillIngesting` problem in `/accounts`.
- New ingestion system is independent of the old one so new endpoint data may not be up to date with the rest of the data in Horizon.

### What shouldn't be reviewed

I plan to add tests and doc updates tomorrow but decided to ask for initial feedback.

### CODE REVIEW

File: services/horizon/internal/actions/account.go [hunk 1]
```diff
@@ -51,3 +54,43 @@ func AccountInfo(ctx context.Context, cq *core.Q, addr string) (*pHorizon.Accoun
 
 	return &resource, errors.Wrap(err, "populating account")
 }
+
+// AccountPage returns a page containing the account records that
+// have a `signer` as a signer.
```
Comment (tamirms) [inline services/horizon/internal/actions/account.go]: I think it should say "have `signer` as a signer"

File: services/horizon/internal/expingest/pipelines.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+		if err != nil {
+			log.
+				WithFields(ilog.F{
+					"ledger": ledgerSeq,
+					"type":   pipelineType,
+					"err":    err,
+				}).
+				Error("Error processing ledger")
+			return historySession.Rollback()
```
Comment (tamirms) [inline services/horizon/internal/expingest/pipelines.go]: I think this should be moved to the top of the function:
```
	p.AddPostProcessingHook(func(ctx context.Context, err error) error {
        	defer historySession.Rollback()
		...
	})
```
It is safe to rollback after a commit

File: services/horizon/internal/expingest/pipelines.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		}
+
+		err = historyQ.UpdateLastLedgerExpIngest(ledgerSeq)
+		if err != nil {
+			return errors.Wrap(err, "Error updating last ingested ledger")
+		}
+
+		// Acquire write lock
+		ingestSession.UpdateLock()
```
Comment (tamirms) [inline services/horizon/internal/expingest/pipelines.go]: do we need to use the ingest session query lock when reading from the accounts for signers table?

Comment (bartekn) [reply services/horizon/internal/expingest/pipelines.go]: It's a copy that's unnecessary from horizon-demo. Just removed it. It's not needed when there's a single data store.

File: services/horizon/CHANGELOG.md [hunk 4]
```diff
@@ -6,9 +6,13 @@ file.  This project adheres to [Semantic Versioning](http://semver.org/).
 As this project is pre 1.0, breaking changes may happen for minor version
 bumps.  A breaking change will get clearly notified in this log.
 
-## v0.18.1
+## v0.19.0
 
 * Add `join` parameter to operations and payments endpoints. Currently, the only valid value for the parameter is `transactions`. If `join=transactions` is included in a request then the response will include a `transaction` field for each operation in the response.
+* Add experimental "Accounts For Signers" endpoint. To enable it set `--enable-accounts-for-signer` CLI param or `ENABLE_ACCOUNTS_FOR_SIGNER=true` env variable. This will exposes `/accounts` endpoint. This requires around 4GB of RAM for initial state ingestion.
```
Comment (ire-and-curses) [inline services/horizon/CHANGELOG.md]: `This will exposes` -> `This will expose a new`

File: services/horizon/internal/db2/history/key_value.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/stellar/go/support/errors"
+)
+
+const (
+	lastLedgerKey = "exp_ingest_last_ledger"
+)
+
+func (q *Q) GetLastLedgerExpIngest() (uint32, error) {
```
Comment (ire-and-curses) [inline services/horizon/internal/db2/history/key_value.go]: Would be good to add docstring

File: services/horizon/internal/db2/history/key_value.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+		ledgerSequence, err := strconv.ParseUint(lastIngestedLedger, 10, 32)
+		if err != nil {
+			return 0, errors.Wrap(err, "Error convering lastIngestedLedger value")
+		}
+
+		return uint32(ledgerSequence), nil
+	}
+}
+
+func (q *Q) UpdateLastLedgerExpIngest(ledgerSequence uint32) error {
```
Comment (ire-and-curses) [inline services/horizon/internal/db2/history/key_value.go]: Docstring

File: services/horizon/internal/db2/history/key_value.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	if err != nil {
+		return 0, err
+	}
+
+	if lastIngestedLedger == "" {
+		return 0, nil
+	} else {
+		ledgerSequence, err := strconv.ParseUint(lastIngestedLedger, 10, 32)
+		if err != nil {
+			return 0, errors.Wrap(err, "Error convering lastIngestedLedger value")
```
Comment (ire-and-curses) [inline services/horizon/internal/db2/history/key_value.go]:
```suggest
return 0, errors.Wrap(err, "error converting lastIngestedLedger value")
```

File: services/horizon/internal/expingest/main.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
+// Package newingest contains the new ingestion system for horizon.
```
Comment (ire-and-curses) [inline services/horizon/internal/expingest/main.go]:
```suggest
 // Package expingest contains the new ingestion system for horizon.
```

File: services/horizon/internal/init.go [hunk 9]
```diff
@@ -66,6 +67,29 @@ func initIngester(app *App) {
...
+
+	if !app.config.EnableAccountsForSigner {
+		return
+	}
+
+	var err error
+	app.expingester, err = expingest.NewSystem(expingest.Config{
+		CoreSession:    app.CoreSession(context.Background()),
+		HistorySession: app.HorizonSession(context.Background()),
+		// Use the first archive for now. We don't have a mechanism to
```
Comment (ire-and-curses) [inline services/horizon/internal/init.go]: Maybe add a TODO here so we don't forget about this

**Pull request merged**